### ISSUE TITLE
Wireless Networking SSID Enumeration Fails to Process Spaces Correctly

### ISSUE THREAD

**Issue by oscilated** [substring/os#6]
Within setup, the wireless networking SSID scan results page incorrectly processes wireless networks that contain spaces.  For instance, if I have a wireless network named "Test Network", the scan results will show two networks, one named "Test" and one named "Network".

**Comment by oscilated** [comment]
If multiple SSIDs with spaces are detected, not only does the results page incorrectly display the network names, but also incorrectly displays the number of networks detected.  For example, if there are two available SSIDs, "Test Network" and "Foo Bar", the scan results will show four networks:

1. Test
2. Network
3. Foo
4. Bar

It's important to also know that if you attempt to join "Foo Bar", you need to choose option 2, despite 2's label of "Network".  As you can imagine, this becomes very confusing if a wireless adapter can detect multiple SSIDs with spaces in the names.

Let me know if you need additional details.

**Comment by substring** [comment]
I had hoped that this would work OOTB since I took all everything required too avoid such traps ... Well, I'll check that and keep you informed, thank you for reporting :)

**Comment by oscilated** [comment]
Anytime.

Let me know if you need additional testing.

For what it's worth, the WiFi setup script in your release correctly handles spaces in the password.  The setup scripts in the December 2020 release of GA did not correctly handle spaces in the password.  But the older versions did handle SSID names correctly...

**Comment by substring** [comment]
rewriting whole parts of gasetup as it was impossible to maintain the way the previous dev(s) wrote it. Not enough PnP, or user oriented, or bad shell scripting etc ... WiFi was the first big step for a cleaner code,  With bugs, obiously 😄

**Comment by substring** [comment]
So I need some info from you as I have no SSID with a space inside. Can you give me the output of `sudo iw dev wlan0 scan | grep "SSID:"` so i can see how it looks like ?

**Comment by oscilated** [comment]
Again, I'm going to post using a quote block because the code block is mangling line feeds:

> [arcade@GroovyArcade ~]$ sudo iw dev wlan0 scan | grep "SSID:"
        SSID: Asteroid Belt
        SSID: rinnai-0021CC31CE52
        SSID: Metis
        SSID: Oort Cloud
        SSID: Kuiper Belt
        SSID: openwireless.org
        SSID: openwireless.org
        SSID: Kuiper Belt
        SSID: Oort Cloud
        SSID: Asteroid Belt
        SSID:
        SSID: Kuiper Belt
        SSID: openwireless.org
        SSID: Oort Cloud

**Comment by substring** [comment]
Mmmmhhhh ... Can you edit /opt/gasetup/core/libs/lib-network.sh and replace [line104](https://gitlab.com/groovyarcade/gasetup/-/blob/master/core/libs/lib-network.sh#L104) be like :

```bash
opts_list="$(IFS=$'\n' ; for a in $list ; do echo -n "$i $a " ; i=$((i + 1)) ; done)"
```

And confirm it lists and sets fine ?

**Comment by oscilated** [comment]
After commenting out the existing
`opts_list=$(for a in $list ; do echo -n "$i $a " ; i=$((i + 1)) ; done)"`  at line 103, adding
`opts_list="$(IFS=$'\n' ; for a in $list ; do echo -n "$i $a " ; i=$((i + 1)) ; done)` on a new line below, the script scans for SSIDs but never displays the results.  Instead it returns to the "Setup Network" menu.

I hope that's what you wanted, because my line 104 is `ask_option no "Select WiFi SSID" "Choose a SSID in the list" required $opts_list`.

**Comment by oscilated** [comment]
This is what my `select_ssid()` function currently looks like:

```
select_ssid() {
  local iface="$1"
  local list=
  inform "Scanning SSIDs on adapter ${iface}..."
  list=$(list_ssid "$iface")
  local i=1
  # opts_list="$(for a in $list ; do echo -n "$i $a " ; i=$((i + 1)) ; done)"
  opts_list="$(IFS=$'\n' ; for a in $list ; do echo -n "$i $a " ; i=$((i + 1)) ; done)"
  ask_option no "Select WiFi SSID" "Choose a SSID in the list" required $opts_list
  [[ $? == 0 ]] || return 1
  echo "$list" | sed -n "${ANSWER_OPTION}p"
  return 0
}
```

**Comment by substring** [comment]
Seems ok with this:
```bash
select_ssid() {
  local iface="$1"
  local list=
  local opts_list=()

  inform "Scanning SSIDs on adapter ${iface}..."
  list=$(list_ssid "$iface")
  local i=1
  IFS=$'\n'
  for a in $list ; do 
    opts_list+=($i)
    opts_list+=("$a")
    i=$((i + 1))
  done
  unset IFS
  ask_option no "Select WiFi SSID" "Choose a SSID in the list" required "${opts_list[@]}"
  [[ $? == 0 ]] || return 1
  echo "$list" | sed -n "${ANSWER_OPTION}p"
  return 0
}
```

**Comment by oscilated** [comment]
I can confirm the new select_ssid() function resolves the original issue.

However, I just noticed a few simple bugs related to configuring a wireless network.

1. The enumeration does not deduplicate SSIDs.  For instance, on my network, I have multiple enterprise WiFi radios that transmit the same SSID/network to enable seamless roaming.  As such, after the network setup script scans for SSIDs, it displays the same SSID multiple times, exactly like `iw dev wlan0 scan | grep "SSID:"` does above.  Not sure if this is an issue for you.  Admittedly very few people will encounter this.

2. After enumerating the networks, and the user selects the network to join, if the user cancels instead of entering the WiFi password, the script returns "Success!" instead of simply cancelling without a message.  Obviously most users will simply enter their password instead of cancel, but just wanted to let you know.

Thanks again for all the bug squashing!  You rock!

**Comment by substring** [comment]
easy to fix that:
* `list_ssid()`: append `| sort | uniq | grep -v '^$'` to the iw line to deduplicate
* `ask_string "Type the password for SSID $ssid"`: add ` || return 1`

I'll add these fixes. Thank you for all your feedback :)

**Comment by oscilated** [comment]
I can confirm the modifications work.

This is my current list_ssid() function:
```list_ssid() {
  local iface="$1"

  sudo iw dev "$iface" scan | grep "SSID:"| sed -E "s/[[:space:]]+SSID: (.*)$/\1/g" | sort | uniq | grep -v '^$'
}
```
And this is my current ask_string:
`ask_string "Type the password for SSID $ssid" || return 1`


I can't currently test whether or not joining a wireless network still works because I'm transferring a significant number of files wirelessly.  However, after the file transfer is finished, I'll double-check we didn't accidentally break something.

Thanks again!

**Comment by substring** [comment]
solved in gasetup 1.7.4, released with os 2021.03
